crates/rattler_solve/src/resolvo/mod.rs

```diff
@@ -694,37 +902,72 @@ impl super::SolverImpl for Solver {
         })?;
 
         // Get the resulting packages from the solver.
-        let required_records = solvables
-            .into_iter()
-            .filter_map(
-                |id| match solver.provider().pool.resolve_solvable(id).record {
-                    SolverPackageRecord::Record(rec) => Some(rec.clone()),
-                    SolverPackageRecord::VirtualPackage(_) => None,
-                },
-            )
-            .collect();
+        let mut record_features: HashMap<RepoDataRecord, Option<Vec<String>>> = HashMap::new();
+
+        for id in solvables {
+            match &solver.provider().pool.resolve_solvable(id).record {
+                SolverPackageRecord::Record(rec) => {
+                    record_features.entry((*rec).clone()).or_insert(None);
+                }
+                SolverPackageRecord::RecordWithFeature(rec, feature) => {
+                    let rec = (*rec).clone();
+                    record_features
+                        .entry(rec)
+                        .or_insert_with(|| Some(Vec::new()))
+                        .get_or_insert_with(Vec::new)
+                        .push(feature.clone());
+                }
+                SolverPackageRecord::VirtualPackage(_) => {}
+            }
+        }
 
-        Ok(required_records)
+        Ok(record_features)
     }
 }
 
-fn parse_match_spec<'a>(
-    pool: &Pool<SolverMatchSpec<'a>>,
-    spec_str: &'a str,
-    parse_match_spec_cache: &mut HashMap<&'a str, VersionSetId>,
-) -> Result<VersionSetId, ParseMatchSpecError> {
-    if let Some(spec_id) = parse_match_spec_cache.get(spec_str) {
-        Ok(*spec_id)
+fn parse_match_spec(
+    pool: &Pool<SolverMatchSpec<'_>, NameType>,
+    spec_str: &str,
+    parse_match_spec_cache: &mut HashMap<String, VersionSetId>,
```

Comment:
```suggestion
    parse_match_spec_cache: &mut HashMap<String, Vec<VersionSetId>>,
```

we can cache all the version sets that are returned by this function. Not just the versionset for the base requirement.

Reply:
Wouldn't be this a bit inefficient if lets say we have `pack[feat1, feat2]` and then `pack[feat2, feat3]` and we are essentially adding the version set for `pack[feat2]` twice. 
I have made the change but let me know if I am missing something.

Reply:
The most important thing to cache is the parsing which is relatively slow. Since the optional features are part of the input string it makes sense to me to cache all the versionsets the result from it. 

The current implementation also doesnt cache the parsing, I think that part should be fixed.

Reply:
Makes sense.
Yes – I just pushed a fix. Let me know what you think!